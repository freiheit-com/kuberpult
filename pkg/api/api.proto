syntax = "proto3";

option go_package = "github.com/freiheit-com/kuberpult/pkg/api";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

package api.v1;

service BatchService {
  rpc ProcessBatch (BatchRequest) returns (BatchResponse) {}
}

message BatchRequest {
  repeated BatchAction actions = 1;
}

message BatchAction {
  oneof action {
    CreateEnvironmentLockRequest create_environment_lock = 1;
    DeleteEnvironmentLockRequest delete_environment_lock = 2;
    CreateEnvironmentApplicationLockRequest create_environment_application_lock = 3;
    DeleteEnvironmentApplicationLockRequest delete_environment_application_lock = 4;
    DeployRequest deploy = 5;
    PrepareUndeployRequest prepare_undeploy = 6;
    UndeployRequest undeploy = 7;
    DeleteEnvironmentFromAppRequest delete_env_from_app = 8;
    CreateEnvironmentRequest create_environment = 9;
    ReleaseTrainRequest release_train = 10;
    CreateReleaseRequest create_release = 11;
    CreateEnvironmentGroupLockRequest create_environment_group_lock = 12;
    DeleteEnvironmentGroupLockRequest delete_environment_group_lock = 13;
  }
}

message BatchResponse {
  repeated BatchResult results = 1;
}

message BatchResult {
  oneof result {
    ReleaseTrainResponse release_train = 10;
    CreateReleaseResponse create_release_response = 11;
  }
}

message CreateEnvironmentLockRequest {
  string environment = 1;
  string lock_id = 2;
  string message = 3;
}

message DeleteEnvironmentLockRequest {
  string environment = 1;
  string lock_id = 2;
}

message CreateEnvironmentGroupLockRequest {
  string environment_group = 1;
  string lock_id = 2;
  string message = 3;
}

message DeleteEnvironmentGroupLockRequest {
  string environment_group = 1;
  string lock_id = 2;
}


message CreateEnvironmentApplicationLockRequest {
  string environment = 1;
  string application = 2;
  string lock_id = 3;
  string message = 4;
}

message DeleteEnvironmentApplicationLockRequest {
  string environment = 1;
  string application = 2;
  string lock_id = 3;
}


message CreateReleaseRequest {
  string environment = 1;
  string application = 2;
  string team = 3;
  // env->yaml
  map<string, string> manifests = 4;
  uint64 version = 5;
  string sourceCommitId = 6;
  string sourceAuthor = 7;
  string sourceMessage = 8;
  string sourceRepoUrl = 9;
  string displayVersion = 10;
}

message CreateReleaseResponse {
}

enum LockBehavior {
  Record = 0;
  Fail = 1;
  Ignore = 2;
}

message DeployRequest {
  string environment = 1;
  string application = 2;
  uint64 version = 3;
  bool ignoreAllLocks = 4 [deprecated = true];
  LockBehavior lockBehavior = 5;
}

message PrepareUndeployRequest {
  string application = 1;
}

message UndeployRequest {
  string application = 1;
}

message DeleteEnvironmentFromAppRequest {
  string application = 1;
  string environment = 2;
}

message ReleaseTrainRequest {
  string target = 1;
  string team = 2;
}

message ReleaseTrainResponse {
  string target = 1;
  string team = 2;
}

message Lock {
  string message = 1;
  string lock_id = 3;
  google.protobuf.Timestamp created_at = 4;
  Actor created_by = 5;
}

message LockedError {
  map<string, Lock> environment_locks = 1;
  map<string, Lock> environment_application_locks = 2;
}

service FrontendConfigService {
  rpc GetConfig (GetFrontendConfigRequest) returns (GetFrontendConfigResponse) {}
}

message GetFrontendConfigRequest {
}

message GetFrontendConfigResponse {
  message ArgoCD {
    string baseUrl = 1;
  }
  message Auth {
    message AzureAuthConfig {
      bool enabled = 1;
      string clientId = 2;
      string tenantId = 3;
      string cloudInstance = 4;
      string redirectURL = 5;
    }
    AzureAuthConfig azureAuth= 1;
  }
  ArgoCD argoCd = 1;
  Auth authConfig = 2;
  string sourceRepoUrl = 3;
  string kuberpultVersion = 4;
  string branch = 5;
}


service OverviewService {
  rpc GetOverview (GetOverviewRequest) returns (GetOverviewResponse) {}
  rpc StreamOverview (GetOverviewRequest) returns (stream GetOverviewResponse) {}
}

message GetOverviewRequest {
  // Retrieve the overview at a certain state of the repository. If it's empty, the latest commit will be used.
  string git_revision = 1;
}

message GetOverviewResponse {
  map<string, Application> applications = 2;
  repeated EnvironmentGroup environmentGroups = 3;
  string git_revision = 4;
}

message EnvironmentGroup {
  string environmentGroupName = 1;
  repeated Environment environments = 2;
  // note that the distanceToUpstream should usually be configured to be the same for all envs in this group, but this is not enforced.
  uint32 distanceToUpstream = 3;
}

enum Priority {
  PROD = 0;
  PRE_PROD = 1;
  UPSTREAM = 2;
  OTHER = 3;
}

message EnvironmentConfig {
  message Upstream {
    optional string  environment = 1;
    optional bool    latest = 2;
  }

  message ArgoCD {
    message SyncWindows {
      string           kind = 1; // "allow" or "deny"
      string           schedule = 2; // crontab format
      string           duration = 3; // duration the window is open (or closed)
      repeated string  applications = 4; // applications names
    }
    message Destination {
      string          name = 1;
      string          server = 2;
      optional string namespace = 3;
      optional string appProjectNamespace = 4;
      optional string applicationNamespace = 5;
    }
    message AccessEntry {
      string group = 1;
      string kind = 2;
    }
    message IgnoreDifferences {
      string          group = 1;
      string          kind = 2;
      string          name = 3;
      string          namespace = 4;
      repeated string jsonPointers = 5;
      repeated string jqPathExpressions = 6;
      repeated string managedFieldsManagers = 7;
    }

    repeated SyncWindows              syncWindows = 1;
    Destination                       destination = 2;
    repeated AccessEntry              accessList = 3;
    map<string, string>               applicationAnnotations = 4;
    repeated IgnoreDifferences        ignoreDifferences = 5;
    repeated string                   syncOptions = 6;
  }

  Upstream upstream = 1;
  ArgoCD argocd  = 2;
  optional string environmentGroup = 3;
}


message CreateEnvironmentRequest {
  string environment = 1;
  EnvironmentConfig config = 2;
}

message Warning {
  oneof warningType {
    UnusualDeploymentOrder unusual_deployment_order = 1;
    UpstreamNotDeployed upstream_not_deployed = 2;
  }
}

message UnusualDeploymentOrder {
  uint64 upstreamVersion = 1;
  string upstreamEnvironment = 2;
  uint64 thisVersion = 3;
  string thisEnvironment = 4;
}

message UpstreamNotDeployed {
  string upstreamEnvironment = 2;
  uint64 thisVersion = 3;
  string thisEnvironment = 4;
}

message Environment {

  message Application {
    message ArgoCD {
      message SyncWindow {
        string kind = 1; // "allow" or "deny"
        string schedule = 2; // crontab format
        string duration = 3; // duration the window is open (or closed)
      }
      repeated SyncWindow syncWindows = 1;
    }
    message DeploymentMetaData {
      string deployAuthor = 1;
      // we use a string here, because the UI cannot handle int64 as a type.
      // the string contains the unix timestamps in seconds (utc)
      string deployTime = 2;
    }

    string name = 1;
    // version=0 means "nothing is deployed"
    uint64 version = 2;
    map<string, Lock> locks = 3;
    // "version" describes the currently deployed version. "queuedVersion" describes a version that was to be deployed, but a lock stopped the deployment:
    // "queuedVersion" has nothing to do with queue.go
    // queuedVersion=0 means "nothing is queued"
    uint64 queuedVersion = 4;
    // google.protobuf.Timestamp deploy_date = 5; // This is never used
    bool undeployVersion = 6;
    ArgoCD argoCD = 7;
    DeploymentMetaData deploymentMetaData = 8;
  }

  string name = 1;
  EnvironmentConfig config = 2;
  map<string, Lock> locks = 3;
  map<string, Application> applications = 4;
  uint32 distanceToUpstream = 5;
  Priority priority = 6;
}

message Release {
  uint64 version = 1;
  string source_commit_id = 2;
  string source_author = 3;
  string source_message = 4;
  google.protobuf.Timestamp created_at = 5;
  bool undeployVersion = 6;
  string prNumber = 7;
  string displayVersion = 8;
}

enum UndeploySummary {
  // "normal": usual case for an active app, there is no undeploy version deployed in any environment
  Normal = 0;
  // "undeploy": all versions are in "undeploy" or don't exist on an environment
  Undeploy = 1;
  // "mixed": undeploy is deployed in one or more, but not all environments
  Mixed = 2;
}

message Application {
  string name = 1;
  repeated Release releases = 2;
  string sourceRepoUrl= 3;
  string team = 4;
  UndeploySummary undeploySummary = 5;
  repeated Warning warnings = 8;
}

message Actor {
  string name = 1;
  string email = 2;
}

service RolloutService {
  rpc StreamStatus (StreamStatusRequest) returns (stream StreamStatusResponse) {}
}

message StreamStatusRequest {}

/*

Unknown = can't tell
Successful = everything is fine
Progressing = waiting for something
Error = error
*/
enum RolloutStatus {
  RolloutStatusUnknown = 0; // argocd didn't report anything for this app
  RolloutStatusSuccesful = 1; // sync healthy
  RolloutStatusProgressing = 2; // out-of-sync || 
  RolloutStatusError = 3; // error in argocd
}

message StreamStatusResponse {
  string environment = 1;
  string application = 2;
  uint64 version     = 3;
  RolloutStatus rollout_status = 4;
}
