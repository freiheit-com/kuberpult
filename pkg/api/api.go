/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
package api

// This file is needed to mark this directory as go package and not fail renovate PRs, the go code here is generally generated by buf and not checked in so renovate thinks this is not a go package
