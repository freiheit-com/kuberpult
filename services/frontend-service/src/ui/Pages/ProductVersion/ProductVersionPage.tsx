/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright freiheit.com*/

import { useGlobalLoadingState } from '../../utils/store';
import { ProductVersion } from '../../components/ProductVersion/ProductVersion';
import { TopAppBar } from '../../components/TopAppBar/TopAppBar';

export const ProductVersionPage: React.FC = () => {
    const element = useGlobalLoadingState();
    if (element) {
        return element;
    }
    return (
        <div>
            <TopAppBar showAppFilter={true} showTeamFilter={true} showWarningFilter={false} showGitSyncStatus={false} />
            <main className="main-content">
                <ProductVersion />
            </main>
        </div>
    );
};
