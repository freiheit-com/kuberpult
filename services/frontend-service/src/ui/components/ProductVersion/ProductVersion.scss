/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright freiheit.com*/
/* The actual timeline (the vertical ruler) */
@import '../../../assets/variables';

.product_version {
    position: relative;
    max-width: 1200px;
    margin: 0 auto;
}

.environment_name {
    font-weight: 700;
    font-size: 28px;
    line-height: 48px;
    padding-left: 35px;
    align-items: center;
    color: var(--mdc-theme-primary);
}

.page_description {
    font-style: normal;
    position: relative;
    font-weight: $small-warn-button-font-weight;
    line-height: $nav-list-item-gap;
    padding: 10px 40px;
    color: var(--mdc-theme-on-surface);
    justify-content: space-between;
    display: flex;
    background: linear-gradient(var(--mdc-theme-primary), var(--mdc-theme-primary)) left 9px top 14px/2px 100% no-repeat;
}

.space_apart_row {
    justify-content: space-between;
    display: flex;
    padding: 20px;
    padding-left: 35px;
}

.release_train_button {
    position: relative;
    margin-left: auto;
    margin-right: $environment-actions-right-displacement;
    margin-top: $environment-actions-top-displacement;
    border: $environment-add-lock-button-border;
    border-radius: $environment-add-lock-button-border-radius;
    color: $environment-add-lock-button-color;
    margin: $environment-buttons-column-margin;
}

.drop_down {
    border-radius: 3px;
    min-width: 220px;
    height: 50px;
    border: 1px solid var(--mdc-theme-primary);
    padding-left: 1em;
    margin-right: 20px;
}

.env_drop_down {
    border-radius: 3px;
    min-width: 220px;
    height: 50px;
    border: 1px solid var(--mdc-theme-primary);
    padding-left: 1em;
    margin-right: 20px;
}

.table_padding {
    padding: 20px;
    padding-left: 35px;
}

.table {
    padding: 10px;
    border-spacing: 0;
    border: 2px solid var(--mdc-theme-primary);
    width: 100%;
    border-radius: 4px;
}

.table_title {
    background-color: var(--mdc-theme-primary);
    color: white;
}
