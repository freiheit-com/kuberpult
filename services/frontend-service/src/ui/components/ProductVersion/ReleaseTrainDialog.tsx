/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import * as React from 'react';
import { useState } from 'react';
import { Checkbox } from '../dropdown/checkbox';
import { ConfirmationDialog } from '../dialog/ConfirmationDialog';
import { useEnvironments } from '../../utils/store';

export type ReleaseTrainDialogProps = {
    environment: string;
    onCancel: () => void;
    open: boolean;
};

export const ReleaseTrainDialog: React.FC<ReleaseTrainDialogProps> = (props) => {
    const [selectedEnvs, setSelectedEnvs] = useState<string[]>([]);

    const envsList = useEnvironments();
    const onConfirm = React.useCallback(() => {
        setSelectedEnvs([]);
    }, []);

    const onCancel = React.useCallback(() => {
        props.onCancel();
        setSelectedEnvs([]);
    }, [props]);

    const addEnv = React.useCallback(
        (env: string) => {
            const newEnv = env;
            const indexOf = selectedEnvs.indexOf(newEnv);
            if (indexOf >= 0) {
                const copy = selectedEnvs.concat();
                copy.splice(indexOf, 1);
                setSelectedEnvs(copy);
            } else {
                setSelectedEnvs(selectedEnvs.concat(newEnv));
            }
        },
        [selectedEnvs]
    );

    return (
        <ConfirmationDialog
            classNames={'env-selection-dialog'}
            onConfirm={onConfirm}
            onCancel={onCancel}
            open={props.open}
            headerLabel={'Select which environments to run release train to:'}
            confirmLabel={'Release Train'}>
            {envsList.filter((env, index) => props.environment === env.config?.upstream?.environment).length > 0 ? (
                <div className="envs-dropdown-select">
                    {envsList
                        .filter((env, index) => props.environment === env.config?.upstream?.environment)
                        .map((env) => {
                            const enabled = selectedEnvs.includes(env.name);
                            return (
                                <div key={env.name}>
                                    <Checkbox
                                        enabled={enabled}
                                        onClick={addEnv}
                                        id={String(env.name)}
                                        label={env.name}
                                        classes={'env' + env.name}
                                    />
                                </div>
                            );
                        })}
                </div>
            ) : (
                <div>There are no environments that have environment {props.environment} as the upstream target</div>
            )}
        </ConfirmationDialog>
    );
};
