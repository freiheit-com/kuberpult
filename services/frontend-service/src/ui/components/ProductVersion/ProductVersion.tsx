/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import './ProductVersion.scss';
import * as React from 'react';
import { refreshTags, useTags, getSummary, useSummaryDisplay } from '../../utils/store';
import { DisplayManifestLink, DisplaySourceLink } from '../../utils/Links';
import { Spinner } from '../Spinner/Spinner';
import { ProductSummary } from '../../../api/api';

export type ProductVersionProps = {
    environment: string;
};

export const ProductVersion: React.FC<ProductVersionProps> = (props) => {
    React.useEffect(() => {
        setShowTagsSpinner(true);
        refreshTags();
    }, []);
    const { environment } = props;
    const summaryResponse = useSummaryDisplay();
    const [open, setOpen] = React.useState(false);
    const openClose = React.useCallback(
        (e: React.ChangeEvent<HTMLSelectElement>) => {
            setShowSummarySpinner(true);
            getSummary(e.target.value, environment);
            setOpen(!open);
        },
        [open, setOpen, environment]
    );
    const [showTagsSpinner, setShowTagsSpinner] = React.useState(false);
    const [showSummarySpinner, setShowSummarySpinner] = React.useState(false);
    var versionToDisplay = (app: ProductSummary): string => {
        if (app.displayVersion !== '') {
            return app.displayVersion;
        }
        if (app.commitId !== '') {
            return app.commitId;
        }
        return app.version;
    };

    const tagsResponse = useTags();
    const [displaySummary, setDisplayVersion] = React.useState(false);

    React.useEffect(() => {
        if (tagsResponse.response.tagData.length > 0) {
            setShowSummarySpinner(true);
            getSummary(tagsResponse.response.tagData[0].commitId, environment);
            setDisplayVersion(true);
        }
    }, [tagsResponse, environment]);
    React.useEffect(() => {
        if (tagsResponse.tagsReady) {
            setShowTagsSpinner(false);
        }
    }, [tagsResponse]);
    React.useEffect(() => {
        if (summaryResponse.summaryReady) {
            setShowSummarySpinner(false);
        }
    }, [summaryResponse]);
    if (showTagsSpinner) {
        return <Spinner message="Loading Tag Data" />;
    }
    if (showSummarySpinner) {
        return <Spinner message="Loading Summary Data" />;
    }

    return (
        <div className="product_version">
            <h1 className="environment_name">{'Product Version for ' + environment}</h1>
            <div className="dropdown_div">
                <select onChange={openClose} onSelect={openClose} className="drop_down" data-testid="drop_down">
                    <option value="default" disabled>
                        Select a Tag
                    </option>
                    {tagsResponse.response.tagData.map((tag) => (
                        <option value={tag.commitId} key={tag.tag}>
                            {tag.tag.slice(10)}
                        </option>
                    ))}
                </select>
            </div>
            <div>
                {displaySummary ? (
                    <div className="table_padding">
                        <table className="table">
                            <tr className="table_title">
                                <th>App Name</th>
                                <th>Version</th>
                                <th>ManifestRepoLink</th>
                                <th>SourceRepoLink</th>
                            </tr>
                            {summaryResponse.response.productSummary.map((sum) => (
                                <tr key={sum.app} className="table_data">
                                    <td>{sum.app}</td>
                                    <td>{versionToDisplay(sum)}</td>
                                    <td>
                                        <DisplayManifestLink
                                            app={sum.app}
                                            version={Number(sum.version)}
                                            displayString="Manifest Link"
                                        />
                                    </td>
                                    <td>
                                        <DisplaySourceLink commitId={sum.commitId} displayString={'Source Link'} />
                                    </td>
                                </tr>
                            ))}
                        </table>
                    </div>
                ) : (
                    <div className="page_description">
                        {
                            'This page shows the version of the product for the selected environment based on tags to the repository. If there are no tags, then no data can be shown.'
                        }
                    </div>
                )}
            </div>
        </div>
    );
};
