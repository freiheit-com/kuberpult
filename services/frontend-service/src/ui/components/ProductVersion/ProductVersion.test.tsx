/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright freiheit.com*/
import { act, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Environment, EnvironmentGroup, Priority, ProductSummary, TagData } from '../../../api/api';
import { ProductVersion, TableFiltered } from './ProductVersion';
import { Spy } from 'spy4js';

const mock_UseEnvGroups = Spy('envGroup');
const mock_UseTags = Spy('Overview');

jest.mock('../../utils/store', () => ({
    refreshTags() {
        return {};
    },
    useEnvironmentGroups() {
        return mock_UseEnvGroups();
    },
    useTags() {
        return mock_UseTags();
    },
    useEnvironments() {
        return [];
    },
    showSnackbarError() {},
}));

jest.mock('../../utils/Links', () => ({
    DisplayManifestLink() {
        return <></>;
    },
    DisplaySourceLink() {
        return <></>;
    },
}));

const mockGetProductSummary = jest.fn();

jest.mock('../../utils/GrpcApi', () => ({
    get useApi() {
        return {
            gitService: () => ({
                GetProductSummary: () => mockGetProductSummary(),
            }),
        };
    },
}));
const sampleEnvsA: Environment[] = [
    {
        name: 'tester',
        locks: {},
        applications: {},
        distanceToUpstream: 0,
        priority: Priority.UPSTREAM,
    },
];

describe('Product Version Data', () => {
    type TestData = {
        name: string;
        environmentName: string;
        environmentGroups: EnvironmentGroup[];
        expectedDropDown: string;
        tags: TagData[];
        productSummary: ProductSummary[];
    };
    const data: TestData[] = [
        {
            name: 'No tags to Display',
            environmentName: 'tester',
            tags: [],
            expectedDropDown: '',
            productSummary: [],
            environmentGroups: [
                {
                    environments: [sampleEnvsA[0]],
                    distanceToUpstream: 1,
                    environmentGroupName: 'g1',
                    priority: Priority.UNRECOGNIZED,
                },
            ],
        },
        {
            name: 'tags to Display with summary',
            environmentName: 'tester',
            tags: [{ commitId: '123', tag: 'refs/tags/dummyTag' }],
            expectedDropDown: 'dummyTag',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
            ],
            environmentGroups: [
                {
                    environments: [sampleEnvsA[0]],
                    distanceToUpstream: 1,
                    environmentGroupName: 'g1',
                    priority: Priority.UNRECOGNIZED,
                },
            ],
        },
        {
            name: 'table to be displayed with multiple rows of data',
            environmentName: 'tester',
            tags: [
                { commitId: '123', tag: 'refs/tags/dummyTag' },
                { commitId: '859', tag: 'refs/tags/dummyTag2' },
            ],
            expectedDropDown: 'dummyTag',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
                { app: 'tester', version: '10', commitId: '4565', displayVersion: '', environment: 'dev', team: '' },
            ],
            environmentGroups: [
                {
                    environments: [sampleEnvsA[0]],
                    distanceToUpstream: 1,
                    environmentGroupName: 'g1',
                    priority: Priority.UNRECOGNIZED,
                },
            ],
        },
    ];

    describe.each(data)(`Displays Product Version Page`, (testCase) => {
        // given
        it(testCase.name, async () => {
            // replicate api calls
            mock_UseEnvGroups.returns(testCase.environmentGroups);
            mock_UseTags.returns({ response: { tagData: testCase.tags }, tagsReady: true });
            mockGetProductSummary.mockResolvedValue({ productSummary: testCase.productSummary });
            render(
                <MemoryRouter>
                    <ProductVersion />
                </MemoryRouter>
            );
            await act(global.nextTick);
            expect(document.body).toMatchSnapshot();
            if (testCase.expectedDropDown !== '') {
                expect(document.querySelector('.drop_down')?.textContent).toContain(testCase.expectedDropDown);
                expect(document.querySelector('.env_drop_down')?.textContent).toContain(testCase.environmentName);
            }

            if (testCase.productSummary.length > 0) {
                expect(document.querySelector('.table')?.textContent).toContain('App Name');
            } else {
                expect(document.querySelector('.page_description')?.textContent).toContain(
                    'This page shows the version'
                );
            }
            const releaseTrainButton = screen.queryByText('Run Release Train');
            if (testCase.tags.length > 0) {
                expect(releaseTrainButton).toBeInTheDocument();
            }
        });
    });
});

describe('Test table filtering', () => {
    type TestData = {
        name: string;
        productSummary: ProductSummary[];
        teams: string[];
        expectedApps: string[];
        filteredApps: string[];
    };
    const data: TestData[] = [
        {
            name: 'no rows to display',
            productSummary: [],
            teams: [],
            expectedApps: [],
            filteredApps: [],
        },
        {
            name: 'no teams to filter out',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
            ],
            teams: [],
            expectedApps: ['testing-app'],
            filteredApps: [],
        },
        {
            name: 'a team to filter out',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
                {
                    app: 'testing-app2',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'others',
                },
            ],
            teams: ['sre-team'],
            expectedApps: ['testing-app'],
            filteredApps: ['testing-app2'],
        },
        {
            name: 'no teams to filter out',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
                {
                    app: 'testing-app 2',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'others',
                },
                {
                    app: 'testing-app 3',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'others',
                },
            ],
            teams: ['sre-team', 'others'],
            expectedApps: ['testing-app', 'testing-app 2', 'testing-app 3'],
            filteredApps: [],
        },
        {
            name: 'bigger example',
            productSummary: [
                {
                    app: 'testing-app',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'sre-team',
                },
                {
                    app: 'testing-app 2',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'others',
                },
                {
                    app: 'testing-app 3',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'others',
                },
                {
                    app: 'testing-app 4',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'another',
                },
                {
                    app: 'testing-app 5',
                    version: '4',
                    commitId: '123',
                    displayVersion: 'v1.2.3',
                    environment: 'dev',
                    team: 'last team',
                },
            ],
            teams: ['sre-team', 'others'],
            expectedApps: ['testing-app', 'testing-app 2', 'testing-app 3'],
            filteredApps: ['testing-app 4', 'testing-app 5'],
        },
    ];
    describe.each(data)(`Displays Product Version Table`, (testCase) => {
        it(testCase.name, () => {
            render(<TableFiltered productSummary={testCase.productSummary} teams={testCase.teams} />);
            expect(document.querySelector('.table')?.textContent).toContain('App Name');
            for (let i = 0; i < testCase.expectedApps.length; i++) {
                expect(document.querySelector('.table')?.textContent).toContain(testCase.expectedApps[i]);
            }
            for (let i = 0; i < testCase.filteredApps.length; i++) {
                expect(screen.queryByText(testCase.filteredApps[i])).not.toBeInTheDocument();
            }
        });
    });
});
