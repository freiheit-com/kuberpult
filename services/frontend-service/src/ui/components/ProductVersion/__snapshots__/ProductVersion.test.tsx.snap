// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Version Data Displays Product Version Page No tags to Display 1`] = `
<body>
  <div>
    <div
      class="product_version"
    >
      <h1
        class="environment_name"
      >
        Product Version for tester
      </h1>
      <div
        class="dropdown_div"
      >
        <select
          class="drop_down"
          data-testid="drop_down"
        >
          <option
            disabled=""
            value="default"
          >
            Select a Tag
          </option>
        </select>
      </div>
      <div>
        <div
          class="page_description"
        >
          This page shows the version of the product for the selected environment based on tags to the repository. If there are no tags, then no data can be shown.
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Product Version Data Displays Product Version Page table to be displayed with multiple rows of data 1`] = `
<body>
  <div>
    <div
      class="product_version"
    >
      <h1
        class="environment_name"
      >
        Product Version for tester2
      </h1>
      <div
        class="dropdown_div"
      >
        <select
          class="drop_down"
          data-testid="drop_down"
        >
          <option
            disabled=""
            value="default"
          >
            Select a Tag
          </option>
          <option
            value="123"
          >
            dummyTag
          </option>
          <option
            value="859"
          >
            dummyTag2
          </option>
        </select>
      </div>
      <div>
        <div
          class="table_padding"
        >
          <table
            class="table"
          >
            <tr
              class="table_title"
            >
              <th>
                App Name
              </th>
              <th>
                Version
              </th>
              <th>
                ManifestRepoLink
              </th>
              <th>
                SourceRepoLink
              </th>
            </tr>
            <tr
              class="table_data"
            >
              <td>
                testing-app
              </td>
              <td>
                v1.2.3
              </td>
              <td />
              <td />
            </tr>
            <tr
              class="table_data"
            >
              <td>
                tester
              </td>
              <td>
                4565
              </td>
              <td />
              <td />
            </tr>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Product Version Data Displays Product Version Page tags to Display with summary 1`] = `
<body>
  <div>
    <div
      class="product_version"
    >
      <h1
        class="environment_name"
      >
        Product Version for tester2
      </h1>
      <div
        class="dropdown_div"
      >
        <select
          class="drop_down"
          data-testid="drop_down"
        >
          <option
            disabled=""
            value="default"
          >
            Select a Tag
          </option>
          <option
            value="123"
          >
            dummyTag
          </option>
        </select>
      </div>
      <div>
        <div
          class="table_padding"
        >
          <table
            class="table"
          >
            <tr
              class="table_title"
            >
              <th>
                App Name
              </th>
              <th>
                Version
              </th>
              <th>
                ManifestRepoLink
              </th>
              <th>
                SourceRepoLink
              </th>
            </tr>
            <tr
              class="table_data"
            >
              <td>
                testing-app
              </td>
              <td>
                v1.2.3
              </td>
              <td />
              <td />
            </tr>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;
